Closes #6921 by adding snapshot testing functionality to plot_gate_map() and plot_coupling_map().

Quant02 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Testing functions for the graphing methods plot_gate_map and plot_coupling_map() were added. Motivated from the issue #6921 .

### Details and comments
1. Added the `filename` parameter to the functions plot_gate_map() and plot_coupling_map() in the module `C:\Users\varun\qiskit-terra\qiskit\visualization\gate_map.py`.
2. Added the testing functions for different use cases of the mapping functions plot_gate_map() and plot_coupling_map() in the module `C:\Users\varun\qiskit-terra\test\ipynb\mpl\graph\test_graph_matplotlib_drawer.py`.

Results of snapshot testing for the functions plot_gate_map() and plot_coupling_map():

![mpl_tester-0 - Jupyter Notebook - Google Chrome 16-09-2021 12_06_38](https://user-images.githubusercontent.com/87556598/133674971-0e4eadd9-ba59-4a78-9bc0-e453489fa165.png)

![snapshot test for coupling map](https://user-images.githubusercontent.com/87556598/132741316-b6d0a9a2-ce8b-45da-8b5b-a6ef3f3bdd32.png)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=7003) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=7003) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=7003) it.</sub>

javabster commented:
Hi @Quant02 you can have a look at `qiskit-terra/test/python/visualization/test_gate_map.py` for examples of how to use a mock backend for your tests 😄

javabster commented:
Also it looks like you have some linting errors, you can run `tox -eblack` to resolve those (linting docs [here](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#style-and-lint))

Quant02 commented:
> Hi @Quant02 you can have a look at `qiskit-terra/test/python/visualization/test_gate_map.py` for examples of how to use a mock backend for your tests 😄

It was helpful. I changed it to a mock backend and now it works fine.

Screenshot of plot_gate_map test:

![gate_map check](https://user-images.githubusercontent.com/87556598/132941050-938b2c61-8b16-4671-ba9e-c7d47c0a7371.png)

Quant02 commented:
> Also it looks like you have some linting errors, you can run `tox -eblack` to resolve those (linting docs [here](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#style-and-lint))

Ah! Using `tox -eblack` made the black commands run successfully and manually edited the code for linting errors. Black and Lint commands ran successfully in bash now. Thanks.

javabster commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +164,28 @@ def test_plot_histogram(self):
+
+        num_qubits = 5
+        qubit_coordinates = [[1, 0], [0, 1], [1, 1], [1, 2], [2, 1]]
+        coupling_map = [[0, 1], [1, 0], [1, 2], [1, 3], [2, 1], [3, 1], [3, 4], [4, 3]]
```
I think you can remove some of these edges (e.g. [0,1] and [1,0] are just redrawing the same line as far as I'm aware)

Quant02 replied:
Removed the duplicate edges. 👍

javabster reviewed (commented):
Thanks for your work on this! Just left a few comments, also could you add a release note please?

Quant02 commented on qiskit/visualization/gate_map.py:
```diff
@@ -24,6 +24,7 @@
 def plot_gate_map(
     backend,
     figsize=None,
+    filename=None,
```
Done!

Quant02 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +164,28 @@ def test_plot_histogram(self):
+
+        self.graph_plot_gate_map(backend=backend, filename="gate_map.png")
+
+    def test_plot_coupling_map(self):
```
Added different use cases by changing parameters and number of qubits for plot_gate_map. As plot_gate_map function calls plot_coupling_map, all the tests for different use cases satisfied by plot_gate_map must be satisfied by plot_coupling_map. Hence, left the testing function for plot_coupling_map unchanged (as it was in the previous commit).

javabster replied:
This is great thanks @Quant02! Can you update the doc strings for the tests so that they don't all say the same thing?

Quant02 replied:
Ah! Missed them. Updated now.

Quant02 commented:
> Thanks for your work on this! Just left a few comments, also could you add a release note please?

I added a `.yaml` file for release notes but what is the ID?

javabster commented:
> > Thanks for your work on this! Just left a few comments, also could you add a release note please?
> 
> I added a `.yaml` file for release notes but what is the ID?

PLease follow the instructions in the [CONTRIBUTING.md](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes) file for how to generate the release notes using reno 😄

Quant02 commented:
> > > Thanks for your work on this! Just left a few comments, also could you add a release note please?
> > 
> > 
> > I added a `.yaml` file for release notes but what is the ID?
> 
> Please follow the instructions in the [CONTRIBUTING.md](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes) file for how to generate the release notes using reno 😄

Thanks. Added release notes. Please review it.

javabster reviewed (dismissed):
LGTM 🚀 Thanks @Quant02!

1ucian0 commented on releasenotes/notes/added-snapshot-tests-for-backend-mapping-functions-5961300e09f05be0.yaml:
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    Added the parameter,  `filename`, to plot_gate_map and plot_coupling_name which 
```
```suggestion
    Added the parameter,  `filename`, to :func:`~qiskit.visualization.gate_map.plot_gate_map` and :func:`~qiskit.visualization.gate_map.plot_coupling_name` which 
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, font_color="#ff00ff", filename="font_color.png")
+
+    def test_plot_coupling_map(self):
+        """for testing the plot_coupling_map"""
```
```suggestion
        """Test plot_coupling_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        )
+
+    def test_font_color(self):
+        """for testing font_color parameter of plot_gate_map"""
```
```suggestion
        """Test font_color parameter of plot_gate_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        )
+
+    def test_line_color(self):
+        """for testing line_color parameter of plot_gate_map"""
```
```suggestion
        """Test line_color parameter of plot_gate_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        )
+
+    def test_qubit_labels(self):
+        """for testing qubit_labels parameter of plot_gate_map"""
```
```suggestion
        """Test qubit_labels parameter of plot_gate_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, qubit_size=38, filename="qubit_size.png")
+
+    def test_qubit_color(self):
+        """for testing qubit_color parameter of plot_gate_map"""
```
```suggestion
        """Test qubit_color parameter of plot_gate_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, figsize=(10, 10), filename="figsize.png")
+
+    def test_qubit_size(self):
+        """for testing qubit_size parameter of plot_gate_map"""
```
```suggestion
        """Test qubit_size parameter of plot_gate_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, filename="65_qubit_gate_map.png")
+
+    def test_figsize(self):
+        """for testing figsize parameter of plot_gate_map"""
```
```suggestion
        """Test figsize parameter of plot_gate_map"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, filename="16_qubit_gate_map.png")
+
+    def test_plot_27_qubit_gate_map(self):
+        """test for plot_gate_map using 27 qubit backend"""
```
```suggestion
        """Test plot_gate_map using 27 qubit backend"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, filename="7_qubit_gate_map.png")
+
+    def test_plot_16_qubit_gate_map(self):
+        """test for plot_gate_map using 16 qubit backend"""
```
```suggestion
        """Test plot_gate_map using 16 qubit backend"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, filename="5_qubit_gate_map.png")
+
+    def test_plot_7_qubit_gate_map(self):
+        """test for plot_gate_map using 7 qubit backend"""
```
```suggestion
        """Test plot_gate_map using 7 qubit backend"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
+        self.graph_plot_gate_map(backend=backend, filename="1_qubit_gate_map.png")
+
+    def test_plot_5_qubit_gate_map(self):
+        """test for plot_gate_map using 5 qubit backend"""
```
```suggestion
        """Test plot_gate_map using 5 qubit backend"""
```

1ucian0 commented on test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py:
```diff
@@ -157,6 +171,122 @@ def test_plot_histogram(self):
         self.graph_count_drawer(counts, filename="histogram.png")
 
+    def test_plot_1_qubit_gate_map(self):
+        """test for plot_gate_map using 1 qubit backend"""
```
```suggestion
        """Test plot_gate_map using 1 qubit backend"""
```

1ucian0 reviewed (commented):
Some small changes.

1ucian0 reviewed (approved):
Thanks a lot and welcome to the community!
